### PR TITLE
3.0: integ-tests: build image tests

### PR DESCRIPTION
**Pull request by yuleiwan**
* Create build image fixture to manage failed imagebuilder stack
* Migrate build image, build image with custom components, build image with wrong version tests
* Add image tag, build tag, custom instance role, custom s3 bucket test

Signed-off-by: Yulei Wang <yuleiwan@amazon.com>

**Please See** [Git Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions)

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/integration-tests/tests/createami/test_createami/test_build_image_wrong_pcluster_version/pcluster.config.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Build:
```
Comment (lukeseawalker) [inline tests/integration-tests/tests/createami/test_createami/test_build_image_wrong_pcluster_version/pcluster.config.yaml:1]: can the file be called differently, like `image.config.yaml`?

File: tests/integration-tests/configs/common/common.yaml [hunk 2]
```diff
@@ -88,6 +88,31 @@ create:
...
+        oss: ["alinux2", "ubuntu2004", "centos7"]
+      - regions: ["us-gov-east-1"]
+        instances: {{ common.INSTANCES_DEFAULT_X86 }}
+        oss: ["ubuntu1804"]
+  test_createami.py::test_build_image_custom_components:
+    dimensions:
+      - regions: ["ap-southeast-2"]
+        instances: {{ common.INSTANCES_DEFAULT_X86 }}
+        oss: ["centos7", "ubuntu1804"]
+      - regions: ["us-east-1"]
```
Comment (lukeseawalker) [inline tests/integration-tests/configs/common/common.yaml:106]: can we leave `["eu-west-1"]` as it was before?

File: tests/integration-tests/conftest.py [hunk 3]
```diff
@@ -908,3 +909,31 @@ def _pcluster_ami_without_standard_naming(version):
...
+    region_post_test = None
+
+    def _build_image(image_name, region, cluster_config):
+        nonlocal image_name_post_test
+        nonlocal region_post_test
+        image_name_post_test = image_name
+        region_post_test = region
+
+        pcluster_build_image_result = run_command(
+            ["pcluster", "build-image", "-n", image_name, "-r", region, "-c", cluster_config.as_posix()]
```
Comment (lukeseawalker) [inline tests/integration-tests/conftest.py:927]: this needs to be updated accordingly to https://github.com/aws/aws-parallelcluster/pull/2711

File: tests/integration-tests/conftest.py [hunk 4]
```diff
@@ -908,3 +909,31 @@ def _pcluster_ami_without_standard_naming(version):
...
+        )
+        return pcluster_build_image_result.stdout
+
+    yield _build_image
+    if image_name_post_test:
+        pcluster_describe_image_result = run_command(
+            ["pcluster", "describe-image", "-n", image_name_post_test, "-r", region_post_test]
```
Comment (lukeseawalker) [inline tests/integration-tests/conftest.py:934]: same here, based on https://github.com/aws/aws-parallelcluster/pull/2711

File: tests/integration-tests/conftest.py [hunk 5]
```diff
@@ -908,3 +909,31 @@ def _pcluster_ami_without_standard_naming(version):
...
+        )
+        logging.info("Build image post process. Describe image result: %s" % pcluster_describe_image_result.stdout)
+
+        if "build_failed" or "error" in pcluster_describe_image_result.stdout.lower():
+            run_command(["pcluster", "delete-image", "-n", image_name_post_test, "-r", region_post_test])
```
Comment (lukeseawalker) [inline tests/integration-tests/conftest.py:939]: same here, based on https://github.com/aws/aws-parallelcluster/pull/2711

Comment (yuleiwan) [reply tests/integration-tests/conftest.py]: done

File: tests/integration-tests/conftest.py (hunk 4 above)
Comment (yuleiwan) [reply tests/integration-tests/conftest.py]: done

File: tests/integration-tests/conftest.py (hunk 3 above)
Comment (yuleiwan) [reply tests/integration-tests/conftest.py]: done

File: tests/integration-tests/configs/common/common.yaml (hunk 2 above)
Comment (yuleiwan) [reply tests/integration-tests/configs/common/common.yaml]: done

File: tests/integration-tests/tests/createami/test_createami.py [hunk 6]
```diff
@@ -11,133 +11,225 @@
...
-    assert_that(os != wrong_os).is_true()
-    base_ami = retrieve_latest_ami(region, wrong_os, ami_type="pcluster", architecture=architecture)
+    while "BUILD_IN_PROGRESS" in pcluster_describe_image_result.stdout:
+        time.sleep(600)
+        pcluster_describe_image_result = run_command(["pcluster", "describe-image", "--id", image_id, "-r", region])
+        logging.info(pcluster_describe_image_result.stdout)
 
-    command = _compose_command(region, instance, os, request, vpc_stack, base_ami, cluster_config)
-    _createami_and_assert_error(command, fr"custom AMI.+{wrong_os}.+base.+os.+config file.+{os}")
+    while "DELETE_IN_PROGRESS" in pcluster_describe_image_result.stdout:
```
Comment (lukeseawalker) [inline tests/integration-tests/tests/createami/test_createami.py:221]: not needed

File: tests/integration-tests/conftest.py [hunk 7]
```diff
@@ -908,3 +909,31 @@ def _pcluster_ami_without_standard_naming(version):
...
+        return pcluster_build_image_result.stdout
+
+    yield _build_image
+    if image_id_post_test:
+        pcluster_describe_image_result = run_command(
+            ["pcluster", "describe-image", "--id", image_id_post_test, "-r", region_post_test]
+        )
+        logging.info("Build image post process. Describe image result: %s" % pcluster_describe_image_result.stdout)
+
+        if "build_failed" or "error" in pcluster_describe_image_result.stdout.lower():
```
Comment (lukeseawalker) [inline tests/integration-tests/conftest.py:938]: can we add a reminder or a TODO that command output parsing needs to be revisited once the command will output a proper JSON?

File: tests/integration-tests/conftest.py [hunk 8]
```diff
@@ -987,3 +987,34 @@ def _pcluster_ami_without_standard_naming(version):
     if ami_id:
         client = boto3.client("ec2", region_name=region)
         client.deregister_image(ImageId=ami_id)
+
+
+@pytest.fixture()
+def build_image():
```
[the remark this answers is not in the document]
Comment (enrico-usai) [reply tests/integration-tests/conftest.py:993]: For now we're just testing image build and image delete but I'll need to add all the logs retrieval commands.

Anyway feel free to merge it and address these changes separately.

File: tests/integration-tests/tests/createami/test_createami.py [hunk 9]
```diff
@@ -11,133 +11,211 @@
...
+):
+    """Test build image for given region and os"""
+    # Get base AMI
+    # remarkable AMIs are not available for ARM and ubuntu2004 yet
+    if os != "ubuntu2004":
+        base_ami = retrieve_latest_ami(region, os, ami_type="remarkable", architecture=architecture)
+    else:
+        base_ami = retrieve_latest_ami(region, os, architecture=architecture)
+
+    image_id = "-".join(["integ", "test", "build", "image", generate_random_string()])
```
Comment (enrico-usai) [inline tests/integration-tests/tests/createami/test_createami.py:44]: minor: an alternative is simply: `image_id = f"integ-test-build-image-{generate_random_string()}"`

File: tests/integration-tests/tests/createami/test_createami.py [hunk 10]
```diff
@@ -11,133 +11,211 @@
...
-    assert_that(stdout_lower).contains("downloading https://{0}-aws-parallelcluster.s3".format(region))
-    assert_that(stdout_lower).does_not_contain("chef.io/chef/install.sh")
-    assert_that(stdout_lower).does_not_contain("packages.chef.io")
-    assert_that(stdout_lower).contains("thank you for installing cinc client")
-    assert_that(stdout_lower).contains("starting cinc client")
-    assert_that(stdout_lower).does_not_contain("no custom ami created")
 
+def _assert_image_tag_and_volume(image_id):
+    logging.info("Check the image tag is present as specified in config file.")
+    image_list = (
```
Comment (enrico-usai) [inline tests/integration-tests/tests/createami/test_createami.py:79]: I think here we have the alternative to use describe-images command to see if it's correctly tagged. In this way we're also testing the describe image command.

File: tests/integration-tests/tests/createami/test_createami.py [hunk 11]
```diff
@@ -11,133 +11,211 @@
...
-    post_install_args = ["--post-install", post_install_script]
-
-    pcluster_createami_result = run_command(
-        ["pcluster", "createami", "-ai", base_ami, "-os", os, "-r", region, "-c", cluster_config.as_posix()]
-        + custom_cookbook_args
-        + instance_args
-        + networking_args
-        + post_install_args,
-        timeout=7200,
+    image_id = "-".join(["integ", "test", "build", "image", "custom", "components", generate_random_string()])
```
Comment (enrico-usai) [inline tests/integration-tests/tests/createami/test_createami.py:185]: same here, you can avoid the join

File: tests/integration-tests/tests/createami/test_createami.py [hunk 12]
```diff
@@ -11,133 +11,211 @@
...
+def _assert_build_image_success(result, image_id, region):
+    logging.info(f"Test build image process for image {image_id}.")
+    assert_that(result).contains("CREATE_IN_PROGRESS")
+    assert_that(result).contains("Build image started successfully.")
 
-@pytest.mark.dimensions("eu-central-1", "c5.xlarge", "ubuntu1804", "*")
-def test_createami_wrong_os(region, instance, os, request, pcluster_config_reader, vpc_stack, architecture):
-    """Test error message when os provide is different from the os of custom AMI"""
-    cluster_config = pcluster_config_reader()
+    pcluster_describe_image_result = run_command(["pcluster", "describe-image", "--id", image_id, "-r", region])
```
Comment (enrico-usai) [inline tests/integration-tests/tests/createami/test_createami.py:204]: here you should call image_factory, we should have something like:
```
image = image_factory(image_config) # in the main test function

image.describe() # here
```

By using this approach we can avoid the duplication of this describe-image command with all the parameter in multiple places and just let the factory manage them.

File: tests/integration-tests/tests/createami/test_createami.py [hunk 13]
```diff
@@ -146,58 +224,30 @@ def test_createami_wrong_pcluster_version(
...
-            "-r",
-            region,
-            "-c",
-            cluster_config.as_posix(),
-            "-i",
-            instance,
-        ]
-        + custom_cookbook_args
-        + networking_args
+    cluster_config = pcluster_config_reader(
```
Comment (enrico-usai) [inline tests/integration-tests/tests/createami/test_createami.py:227]: this should be image_config

File: tests/integration-tests/tests/createami/test_createami.py [hunk 14]
```diff
@@ -11,133 +11,211 @@
...
-
-    pcluster_createami_result = run_command(
-        ["pcluster", "createami", "-ai", base_ami, "-os", os, "-r", region, "-c", cluster_config.as_posix()]
-        + custom_cookbook_args
-        + instance_args
-        + networking_args
-        + post_install_args,
-        timeout=7200,
+    image_id = "-".join(["integ", "test", "build", "image", "custom", "components", generate_random_string()])
+    cluster_config = pcluster_config_reader(
```
Comment (enrico-usai) [inline tests/integration-tests/tests/createami/test_createami.py:186]: this should be image_config

File: tests/integration-tests/tests/createami/test_createami.py [hunk 15]
```diff
@@ -146,58 +224,30 @@ def test_createami_wrong_pcluster_version(
...
+        config_file="image.config.yaml",
+        parent_image=wrong_ami,
+        instance_type=instance,
     )
+    image_id = "-".join(["integ", "test", "build", "image", "wrong", "version", generate_random_string()])
+
+    result = build_image(image_id, region, cluster_config)
+
+    _assert_build_image_failed(result, image_id, region)
+    # TODO get cloudwatch log
```
Comment (enrico-usai) [inline tests/integration-tests/tests/createami/test_createami.py:237]: I have a task for it. I'm creating the commands to retrieve the logs: https://github.com/aws/aws-parallelcluster/pull/2863 

We'll add them in the factory to be used when needed.

File: tests/integration-tests/tests/createami/test_createami/test_build_image_custom_components/image.config.yaml [hunk 16]
```diff
@@ -0,0 +1,11 @@
+Build:
+  InstanceType: {{ instance_type }}
+  ParentImage: {{ parent_image }}
+  Components:
+    {% if region == "us-east-1" %}
```
Comment (enrico-usai) [inline tests/integration-tests/tests/createami/test_createami/test_build_image_custom_components/image.config.yaml:5]: Could you add a comment here? Describing what we're testing in the two cases

File: tests/integration-tests/tests/createami/test_createami.py (hunk 9 above)
Comment (yuleiwan) [reply tests/integration-tests/tests/createami/test_createami.py]: done

File: tests/integration-tests/tests/createami/test_createami.py (hunk 11 above)
Comment (yuleiwan) [reply tests/integration-tests/tests/createami/test_createami.py]: done

File: tests/integration-tests/tests/createami/test_createami.py (hunk 13 above)
Comment (yuleiwan) [reply tests/integration-tests/tests/createami/test_createami.py]: done

File: tests/integration-tests/tests/createami/test_createami/test_build_image_custom_components/image.config.yaml (hunk 16 above)
Comment (yuleiwan) [reply tests/integration-tests/tests/createami/test_createami/test_build_image_custom_components/image.config.yaml]: done

File: tests/integration-tests/conftest.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yuleiwan) [reply tests/integration-tests/conftest.py:994]: Will improve tests and address remaining comment in a sparate PR, including
* image factory
* run describe-images to check tags
* use json format reponse once API is done
* retrieve the cloudwatch logs when build fails

File: tests/integration-tests/tests/createami/test_createami.py (hunk 14 above)
Comment (yuleiwan) [reply tests/integration-tests/tests/createami/test_createami.py]: done

**Review by lukeseawalker — APPROVED** [review]

**Pull request merged**